### PR TITLE
Allow passing `--evaluate` alongside `--list` to evaluate parameter default expressions

### PR DESCRIPTION
**Pull request by epbuennig**
This is an initial draft for adding support to pass `--evaluate` as an option to `--list`  to evaluate expressions in parameter defaults. This PR addresses #1405.

A `justfile` like so:
```
key := "value"

recipe arg=key:
    echo {{ arg }}
```
Should then produce the following output:
```
$ just --evaluate --list
Available recipes:
    recipe arg='key'
```
Possible extensions, like evaluation for `--show` or `--dump` have been left for another PR (as discussed [here](https://github.com/casey/just/issues/1405#issuecomment-1312552643)).

### CODE REVIEW

**Comment by epbuennig** [comment]
This currently works as expected on the repo's `justfile`.

I'm not entirely sure how we should approach passing a mutable evaluator reference to the formatting of a parameter (or later to anything, really). With the current API of `DisplayColor` we'd have to use interior mutability as `FmtEval` in `src/parameter.rs` or some other way. This is obviously hacked in to illustrate the problem.

There's also the issue of `--evalaute` not showing up in the usage anymore, which is why the single commented out test in `test/command.rs` is failing, not sure how to approach that.

File: src/config.rs [hunk 1]
```diff
@@ -113,6 +111,25 @@ mod arg {
   pub(crate) const DUMP_FORMAT_VALUES: &[&str] = &[DUMP_FORMAT_JUST, DUMP_FORMAT_JSON];
 }
 
+mod arg_or_cmd {
```
Comment (epbuennig) [inline src/config.rs]: Not too keen on the naming here, but I figured this should be different to `arg` as it serves 2 purposes now.

File: src/parameter.rs [hunk 2]
```diff
@@ -13,6 +13,43 @@ pub(crate) struct Parameter<'src> {
...
+    }
+    if let Some(prefix) = self.0.kind.prefix() {
+      write!(f, "{}", color.annotation().paint(prefix))?;
+    }
+    write!(f, "{}", color.parameter().paint(self.0.name.lexeme()))?;
+    if let Some(ref default) = self.0.default {
+      write!(
+        f,
+        "={}",
+        color.string().paint(&format!(
```
Comment (epbuennig) [inline src/parameter.rs]: The extra `format!` is mainly because regular strings are green, including their quotes, and I didn't look too far into the coloring module to see if this could be done better.

File: src/parameter.rs [hunk 3]
```diff
@@ -13,6 +13,43 @@ pub(crate) struct Parameter<'src> {
...
+    evaluator: &'eval std::cell::RefCell<Evaluator<'src, 'run>>,
+  ) -> FmtEval<'this, 'eval, 'src, 'run> {
+    FmtEval(self, evaluator)
+  }
+}
+
+pub struct FmtEval<'param, 'eval, 'src, 'run>(
+  &'param Parameter<'src>,
+  &'eval std::cell::RefCell<Evaluator<'src, 'run>>,
+);
```
Comment (epbuennig) [inline src/parameter.rs]: I think even if the evaluator was included in `Wrapper`, it would probably cause major problems from the lifetimes alone.
I tried to use `... -> impl ColorDisplay + 'this`, even with explicit bounds that all other lifetimes outlive `'this`, it would complain about some lifetimes missing from the signature.

Not only that, but I imagine this would complicate the current approach of `Wrapper` which uses a trait object and probably faces similar issues.

File: src/parameter.rs (hunk 2 above)
Comment (epbuennig) [reply src/parameter.rs]: Unfortunately, I selected the wrong range here and GitHub won't let me correct it, referring to L43-L47.

File: src/subcommand.rs [hunk 4]
```diff
@@ -458,9 +494,19 @@ impl Subcommand {
...
-          line_width += UnicodeWidthStr::width(
-            format!(" {}", parameter.color_display(Color::never())).as_str(),
-          );
+          line_width += if let Some(evaluator) = &mut evaluator {
+            UnicodeWidthStr::width(
+              format!(
+                " {}",
+                parameter
+                  .format_eval(evaluator)
+                  .color_display(Color::never())
```
Comment (epbuennig) [inline src/subcommand.rs]: It should also be noted that we may want to keep the evaluated result here to avoid another evaluation later.

File: src/config.rs (hunk 1 above)
Comment (casey) [reply src/config.rs]: On reflection and looking at the code, I think we should avoid overloading the `--evaluate` command flag with a second meaning. It complicates the argument parser and makes the help text weird. How about `--list --eval`?

File: src/parameter.rs [hunk 5]
```diff
@@ -13,6 +13,43 @@ pub(crate) struct Parameter<'src> {
   pub(crate) name: Name<'src>,
 }
 
+impl<'src> Parameter<'src> {
+  pub(crate) fn format_eval<'this, 'eval, 'run>(
```
Comment (casey) [inline src/parameter.rs]: Can we do something like this instead?

```rust
use super::*;

/// A single function parameter
#[derive(PartialEq, Debug, Clone, Serialize)]
pub(crate) struct Parameter<'src, T = Expression<'src>> {
  /// An optional default expression
  pub(crate) default: Option<T>,
  /// Export parameter as environment variable
  pub(crate) export: bool,
  /// The kind of parameter
  pub(crate) kind: ParameterKind,
  /// The parameter name
  pub(crate) name: Name<'src>,
}

impl<'src, T: Display> ColorDisplay for Parameter<'src, T> {
  fn fmt(&self, f: &mut Formatter, color: Color) -> Result<(), fmt::Error> {
    if self.export {
      write!(f, "$")?;
    }
    if let Some(prefix) = self.kind.prefix() {
      write!(f, "{}", color.annotation().paint(prefix))?;
    }
    write!(f, "{}", color.parameter().paint(self.name.lexeme()))?;
    if let Some(ref default) = self.default {
      write!(f, "={}", color.string().paint(&default.to_string()))?;
    }
    Ok(())
  }
}
```

This gives Parameter a generic parameter T which is the type of the default. T defaults to `Expression<'src>`, so everything works as normal. You can create an evaluated parameter with `Parameter<String>`.

So when `--eval` is passed, convert parameters to `Parameter<String>` using eval before printing.

File: src/parameter.rs [hunk 6]
```diff
@@ -13,6 +13,43 @@ pub(crate) struct Parameter<'src> {
...
+    FmtEval(self, evaluator)
+  }
+}
+
+pub struct FmtEval<'param, 'eval, 'src, 'run>(
+  &'param Parameter<'src>,
+  &'eval std::cell::RefCell<Evaluator<'src, 'run>>,
+);
+
+impl<'this, 'eval, 'src, 'run> ColorDisplay for FmtEval<'this, 'eval, 'src, 'run> {
```
Comment (casey) [inline src/parameter.rs]: I think that combining fmt and eval is probably the wrong way to go. See my comment above.

**Review by casey — APPROVED** [review]
Thanks for tackling this! Added a comment with an alternative approach that might be much cleaner. Let me know what you think.

**Review by casey — CHANGES_REQUESTED** [review]
Whoops, meant to request changes.

File: src/parameter.rs (hunk 6 above)
Comment (epbuennig) [reply src/parameter.rs]: Yeah, that's what I thought too.

File: src/parameter.rs (hunk 5 above)
Comment (epbuennig) [reply src/parameter.rs]: If `Parameter` becomes generic then when we evaluate it we need to store it separately to the originals because the recipes expect `Parameter<Expression<'src>>`, right? That would add some complexity to simply printing the recipes.
What if we instead evaluate them in place into `Expression::String`? That way the type stays the same, and we can keep most of the formatting code that was already in place, I think.

EDIT:
Ah, I realize now that the only suitable expression type is `StringLiteral` which borrows for `'src`, I guess that won't work. But I'd still like to avoid having to pack up the parameters separately to the recipes when evaluating and checking the length. Would you mind if I instead added an `enum` containing either an Expression or an evaluated expression? To avoid the type of `Parameter` being different after evaluation. Although that would require mutable access.

Comment (epbuennig) [reply src/parameter.rs]: I agree on the flag, I want to avoid changing something like the CLI in a breaking manner. Although, it feels a bit weird to have both `evaluate` and `eval`, do you think we can differentiate them more?

**Comment by epbuennig** [comment]
I'll get to it in the next few days and should have it ready and cleaned up for another review.

**Comment by epbuennig** [comment]
At the risk of this becoming stale, I've been stuck on a peculiar borrowing problem, I'll push the code as is, it does not currently compile.

File: src/subcommand.rs [hunk 7]
```diff
@@ -493,19 +485,12 @@ impl Subcommand {
...
-                parameter
-                  .format_eval(evaluator)
-                  .color_display(Color::never())
-              )
-              .as_str(),
-            )
-          } else {
-            UnicodeWidthStr::width(format!(" {}", parameter.color_display(Color::never())).as_str())
+          if let Some(evaluator) = &mut evaluator {
+            let parameter = parameter.evaluate_default(evaluator)?;
```
Comment (epbuennig) [inline src/subcommand.rs]: According to the borrow checker, we're borrowing from `&justfile.recipes` in line 479 when we return in the `Err` case here. Despite the only lifetime being mentioned in the return value of `evalaute_default` being `'src`. I'm not sure why this is, as it works fine in `justfile.run()`.

Comment (epbuennig) [reply src/subcommand.rs]: Changing the `justfile` parameter of `list` to not be moved but borrowed, yields a similar result, requiring the borrow of `&Justfile<'scr>` itself to outlive `'src`, i.e., suggesting `&'src Justfile<'src>`, which is not at all satiable.

**Comment by epbuennig** [comment]
So I followed your approach first to see where it leads me, other than the borrowing problem I'm stumped by, there are a few other minor problems I want to highlight:

If we evaluate in the nested loop as seen in the current state of this draft, we'll evaluate a parameter once for every alias of a command. Later in the actual formatting step we need to either do this again, doubling all evaluations, or tracking which recipe they belong to, requiring some local of type `BTreeMap<&'src str, Vec<Parameter<'src, String>>` or similar.
I personally would rather mutate the parameters in place, caching the evaluated `String` alongside its original `Expression`. Allowing us to evaluate once and once only, but the `Recipe`s are behind `Rc`s because of the aliases.

File: src/subcommand.rs [hunk 8]
```diff
@@ -485,12 +487,37 @@ impl Subcommand {
...
-          line_width += UnicodeWidthStr::width(
-            format!(" {}", parameter.color_display(Color::never())).as_str(),
-          );
+          // FIXME: unwrap -> ?
+          if let Some(evaluated) = evaluator
+            .as_mut()
+            .map(|evaluator| parameter.evaluate_default(evaluator))
+            .transpose()
+            .unwrap()
+            .flatten()
```
Comment (epbuennig) [inline src/subcommand.rs]: The `unwrap` on L495 is a result of what I previously wrote, I have no idea why it thinks that we're borrowing from the locally owned `justfile` when returning an eval error.
So I implemented the functionality anyway by unwrapping and left the `FIXME` in there.

File: src/subcommand.rs [hunk 9]
```diff
@@ -485,12 +487,37 @@ impl Subcommand {
...
+          {
+            use std::collections::btree_map::Entry;
+
+            line_width += UnicodeWidthStr::width(
+              format!(" {}", evaluated.color_display(Color::never())).as_str(),
+            );
```
Comment (epbuennig) [inline src/subcommand.rs]: We could technically skip this string alloc by simply counting the evaluated and name length + 3, but optimizations are not the point of this PR. I want to keep the diff minimal.

File: src/subcommand.rs [hunk 10]
```diff
@@ -485,12 +487,37 @@ impl Subcommand {
...
+                let mut parameters = BTreeMap::new();
+                parameters.insert(evaluated.name.lexeme(), evaluated);
+                vacant.insert(parameters);
+              }
+              Entry::Occupied(mut occopied) => {
+                occopied
+                  .get_mut()
+                  .insert(evaluated.name.lexeme(), evaluated);
+              }
+            }
```
Comment (epbuennig) [inline src/subcommand.rs]: I've tested this locally with a justfile like this:
```make
key1 := `echo val11`
key2 := `echo val22`

recipe1 arg1=key1 arg2=`echo val12`:
    echo

recipe2 arg1=`echo val21` arg2=key2:
    echo
```
But I wasn't sure how to exactly write a test here to fit in with the rest, if that is even necessary.

File: src/config.rs [hunk 11]
```diff
@@ -522,11 +529,15 @@ impl Config {
...
     } else if matches.is_present(cmd::LIST) {
-      Subcommand::List
+      Subcommand::List {
+        evaluate: matches.is_present(arg::EVAL),
+      }
     } else if let Some(name) = matches.value_of(cmd::SHOW) {
       Subcommand::Show {
         name: name.to_owned(),
       }
+    } else if matches.is_present(cmd::VARIABLES) {
```
Comment (casey) [inline src/config.rs]: This should be moved back to the original location, to avoid churn in the diff.

File: src/config.rs [hunk 12]
```diff
@@ -155,6 +156,12 @@ impl Config {
           .value_name("FORMAT")
           .help("Dump justfile as <FORMAT>"),
       )
+      .arg(
+        Arg::with_name(arg::EVAL)
+          .long("eval")
+          .help("Evaluate parameter default expressions.")
```
Comment (casey) [inline src/config.rs]:
```suggestion
          .help("Evaluate parameter defaults.")
```

File: src/subcommand.rs [hunk 13]
```diff
@@ -428,7 +432,35 @@ impl Subcommand {
...
   }
 
-  fn list(config: &Config, justfile: Justfile) {
+  fn list<'src>(
+    config: &Config,
+    evaluate: bool,
+    search: &Search,
+    justfile: Justfile<'src>,
+  ) -> Result<(), Error<'src>> {
+    let dotenv = if config.load_dotenv {
```
Comment (casey) [inline src/subcommand.rs]: A dotenv is only needed if we're evaluating parameter defaults, so we shouldn't load it if `evaluate` is false.

File: src/subcommand.rs (hunk 8 above)
Comment (casey) [reply src/subcommand.rs]: I just pushed a diff that avoids this. `justfile.recipes.iter()` works for some reason.

File: src/subcommand.rs (hunk 10 above)
Comment (casey) [reply src/subcommand.rs]: What is this testing?

**Review by casey — CHANGES_REQUESTED** [review]
A few minor comments.

I'd like to avoid entangling evaluation with formatting. Check out this PR for a sketch of how to do that:

https://github.com/casey/just/commit/1700cd7d42843994c2676233d77075b606baf4fd

It splits out the formatting into `list_inner`, which should probably be called` list_evaluated`. This should make it easier to do evaluation in one place, and then list_evaluated can be the same regardless of whether it's getting evaluated or unevaluated parameters.

**Comment by epbuennig** [comment]
> What is this testing?

This was simply bad code and is fixed in ff1b69f.

File: src/config.rs (hunk 11 above)
Comment (epbuennig) [reply src/config.rs]: Resolved in 17112ff!

File: src/config.rs (hunk 12 above)
Comment (epbuennig) [reply src/config.rs]: Resolved in d3dfffa!

File: src/subcommand.rs (hunk 13 above)
Comment (epbuennig) [reply src/subcommand.rs]: Resolved in 154bf63!

**Comment by epbuennig** [comment]
I think I've resolved the remaining comments, it currently tests fine on my machine, passing `--eval` evaluates as expected. I'll take a look at whatever obvious improvements to the code can be done and looking for unhandled edge cases before requesting review.

**Comment by epbuennig** [comment]
It's been a while, but I didn't see anything while checking the places I edited. I'm going to request the final review, thank you for being so patient with me. The remaining conflict is simply an inlined format string.
